homeassistant/components/rainmachine/switch.py

```diff
@@ -390,6 +394,14 @@ async def async_turn_on_when_active(self, **kwargs: Any) -> None:
             )
         )
 
+        # When the zone starts, calculate when it is expected to be done and set the
+        # appropriate attribute once:
+        data = self.coordinator.data[self.entity_description.uid]
+        self._attr_extra_state_attributes[ATTR_ZONE_FINISHED_AT] = utcnow() + timedelta(
```

Comment:
Shouldn't this be a sensor with a device class `timestamp` ? 

For Shelly, we calculate the timestamp on every update based on some remaining value from the API, and we only update the value in HA if it's >5s different to avoid updating many times when the value goes faster/slower than HA.